FoundationDB/fdb-kubernetes-operator
# Replace pod for instances with changed file security context

# Description

Should solve https://github.com/FoundationDB/fdb-kubernetes-operator/issues/208 by adding a check to replacement workflow.  It also adds a feature flag (default off) and an e2e test.

## Type of change

- New feature (non-breaking change which adds functionality)


## Discussion

*Are there any design details that you would like to discuss further?*

## Testing

*Please describe the tests that you ran to verify your changes. Unit tests?
Manual testing?*
Unit test, e2e tests
e2e test output (focus tested on the new one)
```
 {"level":"info","ts":"2024-05-21T17:54:19Z","logger":"controller","msg":"Attempting to run sub-reconciler","namespace":"nicmorales","cluster":"nicdev","reconciler":"controllers.replaceMisconfiguredProcessGroups"}
  {"level":"info","ts":"2024-05-21T17:54:19Z","logger":"controller","msg":"Replace process group","namespace":"nicmorales","cluster":"nicdev","reconciler":"controllers.replaceMisconfiguredProcessGroups","reason":"either FSGroup or FSGroupChangePolicy has changed for the pod SecurityContext"}
  {"level":"info","ts":"2024-05-21T17:54:19Z","logger":"controller","msg":"Replace process group","namespace":"nicmorales","cluster":"nicdev","reconciler":"controllers.replaceMisconfiguredProcessGroups","reason":"either FSGroup or FSGroupChangePolicy has changed for the pod SecurityContext"}
  {"level":"info","ts":"2024-05-21T17:54:19Z","logger":"controller","msg":"Replace process group","namespace":"nicmorales","cluster":"nicdev","reconciler":"controllers.replaceMisconfiguredProcessGroups","reason":"either FSGroup or FSGroupChangePolicy has changed for the pod SecurityContext"}
  {"level":"info","ts":"2024-05-21T17:54:19Z","logger":"controller","msg":"Replace process group","namespace":"nicmorales","cluster":"nicdev","reconciler":"controllers.replaceMisconfiguredProcessGroups","reason":"either FSGroup or FSGroupChangePolicy has changed for the pod SecurityContext"}
  {"level":"info","ts":"2024-05-21T17:54:19Z","logger":"controller","msg":"Replace process group","namespace":"nicmorales","cluster":"nicdev","reconciler":"controllers.replaceMisconfiguredProcessGroups","reason":"either FSGroup or FSGroupChangePolicy has changed for the pod SecurityContext"}
  {"level":"info","ts":"2024-05-21T17:54:19Z","logger":"controller","msg":"Removals have been updated in the cluster status","namespace":"nicmorales","cluster":"nicdev","reconciler":"controllers.replaceMisconfiguredProcessGroups"}
  {"level":"info","ts":"2024-05-21T17:54:19Z","logger":"controller","msg":"Subreconciler finished run","namespace":"nicmorales","cluster":"nicdev","reconciler":"controllers.replaceMisconfiguredProcessGroups","duration_seconds":0.031225991}
```

*Do we need to perform additional testing once this is merged, or perform in a larger testing environment?*
Any unique kubernetes environment should be checked for securityContext server-side config changes before enabling this feature.

## Documentation

*Did you update relevant documentation within this repository?*
I think so! Added docstrings + section in `docs/manual/replacements_and_deletions.md`.

*If this change is adding new functionality, do we need to describe it in our user manual?*
I think the addition in `docs/manual/replacements_and_deletions.md` covers it but let me know if not.

*If this change is adding or removing subreconcilers, have we updated the core technical design doc to reflect that?*
N/A

*If this change is adding new safety checks or new potential failure modes, have we documented and how to debug potential issues?*
I think they are straightforward enough, but let me know if not

## Follow-up

*Are there any follow-up issues that we should pursue in the future?*

*Does this introduce new defaults that we should re-evaluate in the future?*
